api/src/services/token-login.js

```diff
@@ -0,0 +1,375 @@
+const _ = require('lodash');
+const passwordTester = require('simple-password-tester');
+const db = require('../db');
+const config = require('../config');
+const taskUtils = require('@medic/task-utils');
+const phoneNumber = require('@medic/phone-number');
+const TOKEN_EXPIRE_TIME = 24 * 60 * 60 * 1000; // 24 hours
+const PASSWORD_MINIMUM_LENGTH = 20;
+const PASSWORD_MINIMUM_SCORE = 50;
+const TOKEN_LENGTH = 64;
+
+/**
+ * Generates a complex enough password with a given length
+ * @param {Number} length
+ * @returns {string} - the generated password
+ */
+const generatePassword = (length = PASSWORD_MINIMUM_LENGTH) => {
+  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.,:!$-=';
+  let password;
+  do {
+    password = Array.from({ length }).map(() => _.sample(chars)).join('');
+  } while (passwordTester(password) < PASSWORD_MINIMUM_SCORE);
+
+  return password;
+};
+
+/**
+ * Given user data, validates whether the phone field is filled with a valid phone number.
+ * Throws an error if the phone number field is empty or value is not valid.
+ * If valid, normalizes the field value.
+ * @param {Object} data - the user data
+ * @param {String} data.phone - the phone number to validate
+ */
+const validateAndNormalizePhone = (data) => {
+  const settings = config.get();
+  if (!phoneNumber.validate(settings, data.phone)) {
+    return {
+      msg: 'A valid phone number is required for SMS login.',
+      key: 'configuration.enable.token.login.phone'
+    };
+  }
+
+  data.phone = phoneNumber.normalize(settings, data.phone);
+};
+
+/**
+ * Generates a unique 64 character token.
+ * @returns {String}
+ */
+const generateToken = (retry = 0) => {
+  const maxRetry = 10;
+  if (retry >= maxRetry) {
+    throw new Error('Failed to generate unique token after 10 iterations');
+  }
+  const tokens = Array.from({ length: 10 }).map(() => generatePassword(TOKEN_LENGTH));
+  const docIds = tokens.map(token => getTokenLoginDocId(token));
+  return db.medic.allDocs({ keys: docIds }).then(results => {
+    const idx = results.rows.findIndex(row => row.error);
+    if (idx === -1) {
+      return generateToken(retry + 1);
```

Comment:
I don't think we need to retry - if the first 10 are dupes then give up right away.

Reply:
Done!